File: Python/oop_python_exercises/BankAccount.py
"""Exercise 1:
Create a Python class called "BankAccount" that represents a bank account. 
The class should have attributes for the account number, account holder name, and balance. 
Implement methods to deposit money, withdraw money, and check the account balance. to make it a bit challanging allow 
for the bankaccount to hold multiple accounts.
"""
class BankAccount:
    def __init__(self, account_name, account_number, account_balance):
        self.Bank_info = [{"account_name":account_name, "account_number":account_number, "account_balance":account_balance}]
    

    def add_another_account(self, account_name, account_number, account_balance):
        self.Bank_info.append({"account_name":account_name, "account_number":account_number, "account_balance":account_balance})

    def deposit(self, account_number, amount):
        for i in self.Bank_info:
            if i["account_number"] == account_number:
                i["account_balance"]+=amount
                return i
            
    def withdraw(self, account_number, amount):
        for i in self.Bank_info:
            if i["account_number"] == account_number:
                if i["account_balance"] >= amount:
                    i["account_balance"]-=amount
                    return "Your remaining balance is ", i["account_balance"]
                else:
                    return "Not enough fund"
            
    def account_balance(self):
        print(self.Bank_info)

    
        

bank_accounts = []

x = BankAccount("John", 100,  1000)
x.add_another_account("Mary", 200, 1000)
x.add_another_account("rinvert", 300, 1000)


print(x.account_balance())

print(x.deposit(300, 500))

print(x.account_balance())

print(x.withdraw(200, 500))

print(x.account_balance())

File: Java_learn/exercises/src/main/java/com/java_exercises/BankAccount.java
package com.java_exercises;

/*Exercise: 

Implement a banking application that models different types of bank accounts. 
Each account should be able to perform basic operations such as deposit, withdraw, and display account information. 
Additionally, certain types of accounts should provide specialized functionality.

Start by creating an abstract class BankAccount that encapsulates the common properties and methods for all bank accounts.

The class should have the following:
Instance variables: accountNumber (String) and balance (double).
Abstract methods: deposit(double amount), withdraw(double amount).
Implement the BankAccount class to provide basic functionality for account information display.
Create an interface called InterestBearing that declares the abstract method calculateInterest().
This interface will be implemented by specific types of accounts that earn interest.

Create a concrete class called SavingsAccount that extends BankAccount and implements the InterestBearing interface.
This class should have an additional instance variable interestRate (double) 
and an implementation of the calculateInterest() method specific to savings accounts.

Create a concrete class called CheckingAccount that extends BankAccount. 
This class should have an additional instance variable transactionFee (double) 
and an implementation of the withdraw() method that deducts the transaction fee from the balance.

In the main program, 
you can demonstrate polymorphism by creating a list of BankAccount objects and storing instances of both SavingsAccount and CheckingAccount in the same list. 
You can then iterate over the list and invoke common methods, such as deposit(), withdraw(), and displayAccountInfo(), on each account without knowing the specific account type. */
abstract class BankAccount {
    private String accountNumber;
    private double balance;

    public BankAccount(String accountNumber, double balance){
        this.accountNumber = accountNumber;
        this.balance = balance;
    }

    public abstract double deposit(double amount);
    public abstract void withdraw(double amount);
    public void displayAccountInfo(){
        System.out.println("Account Number: " + accountNumber);
        System.out.println("Balance: $" + balance);
    };


    public String getAccountNumber(){
        return accountNumber;
    }
    public void setAccountNumber(String value){
        accountNumber = value;
    }

    public double getBalance(){
        return balance;
    }
    
    public double setBalance(double amount){
        balance = amount;
        return balance;
    }

    public interface InterestBearing {
        void calculateInterest();
    }
    
}



